promql/engine.go

```diff
@@ -1633,6 +1633,33 @@ func (ev *evaluator) evalSubquery(ctx context.Context, subq *parser.SubqueryExpr
 	return ms, mat.TotalSamples(), ws
 }
 
+// evalBinaryExpr evaluates the binary expression and returns the evaluated
+// MatrixSelector in its place. Note that the Name and LabelMatchers are not set.
+func (ev *evaluator) evalBinaryExpr(ctx context.Context, binexp *parser.BinaryExpr) (*parser.MatrixSelector, int, annotations.Annotations) {
+	val, ws := ev.eval(ctx, binexp)
+	mat := val.(Matrix)
+
+	var ms *parser.MatrixSelector
+	var vs *parser.VectorSelector
+
+	parser.Inspect(binexp, func(node parser.Node, _ []parser.Node) error {
+		if n, ok := node.(*parser.MatrixSelector); ok {
+			ms = n
+		}
+		return nil
+	})
+
+	vs = ms.VectorSelector.(*parser.VectorSelector)
+	vs.Series = make([]storage.Series, 0, len(mat))
```

Comment:
Since there's no conditional appending here and it's a performance-sensitive piece of code, wouldn't it be faster to pre-set the slice length via `make([]storage.Series, len(mat))` and then use `vs.Series[i] = NewStorageSeries(s)` instead of using `append()`? Although I see that the equivalent code in `evalSubquery()` also doesn't do that currently.